### DISCUSSION TITLE
Support reading paths from stdin

### DISCUSSION THREAD

**Discussion by sayanarijit** [claudiodangelis/qr-filetransfer#233] [Ideas]
I'm opening this issue because:

- [ ] I have found a bug
- [x] I want to request a feature
- [ ] I have a question
- [ ] Other

Hi again, is there any possibility of supporting reading file paths from stdin like [pipe-rename](https://github.com/marcusbuffett/pipe-rename)?

This will solve two main problems:

- I won't have to deal with [escaping file paths](https://github.com/sayanarijit/xplr/wiki/Hacks#sendreceive-files-via-qr-code-on-lan) that include white spaces. I can just use
   ```
   cat $XPLR_PIPE_RESULT_OUT | qrcp send -
   ```
- The input will not be limited to the maximum size of command-line arguments supported by the system.

**Comment by claudiodangelis** [comment]
Oh, very good suggestion indeed. I also want to add something like

```
# Say that you want to transfer the output of some command
some-command | qrcp
```

**Comment by claudiodangelis** [comment]
Hello, would something like this be an option?

```sh
qrcp send $(echo $XPLR_PIPE_RESULT_OUT)
```

EDIT: this way it should already work with the current version.

**Comment by sayanarijit** [comment]
> Hello, would something like this be an option?
> 
> ```shell
> qrcp send $(echo $XPLR_PIPE_RESULT_OUT)
> ```
> 
> EDIT: this way it should already work with the current version.

Actually this is what I'm currently using. But when there are multiple inputs, this method has two issues I have listed.

- Having to properly quote/escape file paths with white spaces manually. Some tools allow null character `\0` as a separator. But it's much better to be able to pass newline separated paths via stdin.
- The number of commandline arguments that can be passed are limited and the limit is quite low on some systems. Normally other commandline tools can use something like xargs to workaround this limitation, but it won't work with qrcp (unless we're fine with repeating one shot download for every file).

**Comment by claudiodangelis** [comment]
Ok thanks, can you please paste here a sample content of `XPLR_PIPE_RESULT_OUT` containing multiple inputs?
I will use it as a test string to implement a solution.

Thanks!
C

**Comment by sayanarijit** [comment]
The value of `$XPLR_PIPE_RESULT_OUT` is the path to a text file that contains one or more newline delimited file paths that are selected (to be sent via qrcp).

So if you `cat $XPLR_PIPE_RESULT_OUT`, you get:

```
/path/to/file1
/path/to/file-2
/path/to/file 3
```

And so on.

So, I'm currently using something similar to `qrcp send $(cat $XPLR_PIPE_RESULT_OUT)`. ([See here](https://github.com/sayanarijit/qrcp.xplr/blob/d8cf4f9b57ae10b59c213eb46c5ff5aed793b3ac/init.lua#L28)). Which won't work if files with whitespaces are selected.

**Comment by claudiodangelis** [comment]
Ah ok, now I understand. What about a command line flag to read the list of files to be transferred from a file?
You would use it like:

```
qrcp send --from-file $XPLR_PIPE_RESULT_OUT
```

**Comment by sayanarijit** [comment]
Yes... that definitely works for this use case. IMO, reading from `stdin` (at-least by default) is more flexible, as it will make integrating qrcp with many CLI tools (other than xplr) easier.

For e.g. in Linux, with the `--from-file` option, one can do:

```
find . -name "*.go" | qrcp send --from-file /dev/stdin
```

By using `-` as a shortcut to `--from-file /dev/stdin` it'd look like:

```
find . -name "*.go" | qrcp send -
```

But the second approach allows us to abstract away the path to stdin, which will come in handy in Windows, as there's [apparently no equivalent to /dev/stdin](https://stackoverflow.com/a/7396793/5209439).

**Comment by sayanarijit** [comment]
It may also be possible to use `- <PATH>` as a direct shortcut to `--from-file <PATH>` and if `<PATH>` is not given, the lines will be read from the stdin.

**Comment by claudiodangelis** [comment]
Ok, I like everything you propose.
Let's summarize:
| Command                        | Outcome                                                                         |
|--------------------------------|---------------------------------------------------------------------------------|
| `tail -n 10 logfile \| qrcp`   | The output of the command is transferred as a text file                         |
| `find \| qrcp -`               | The output of the command is considered as a list of files to be transferred    |
| `qrcp --from-file my-list.txt` | The content of `my-list.txt` is considered as a list of files to be transferred |
| `cat my-list.txt \| qrcp -`    | Equivalent to `qrcp --from-file my-list.txt`                                    |

Ok?

**Reply by sayanarijit** [reply]
Almost, but I'm not sure about the first one. It might be confusing for qrcp to detect whether the given lines are files to send or text to expose.
Since the regular `cp` command only deals with whole files and there's no way to copy part of a file or plain text using `cp`, I assumed `qrcp` should have similar features and limitations.
I agree being able to send partial content of a file might be useful. But then again, there are other tools like `qrencode` and https://qrcode.show to do the same.
So it's your call.
Rest of what you posted looks fine. Example 2 and 4 are similar. I think we can combine example 3 with 2/4 by declaring `-` as an alias to `--from-file` which defaults to the `stdin`. Which means `qrcp - /path/to/file` and `qrcp --from-file /path/to/file` are the same. When `/path/to/file` is not given, qrcp will read from stdin.

**Reply by 097115** [reply]
As of now, `qrcp` still doesn't support reading from stdin, right?
